aptos-move/framework/aptos-framework/sources/stake.move

```diff
@@ -981,19 +1014,43 @@ module aptos_framework::stake {
 
     /// Triggers at epoch boundary. This function shouldn't abort.
     ///
-    /// 1. Distribute rewards to stake pools of active and pending inactive validators (requested to leave but not yet
+    /// 1. Distribute transaction fees to stake pools of active and pending inactive validators.
+    /// 2. Distribute rewards to stake pools of active and pending inactive validators (requested to leave but not yet
     /// removed).
-    /// 2. Officially move pending active stake to active and move pending inactive stake to inactive.
+    /// 3. Officially move pending active stake to active and move pending inactive stake to inactive.
     /// The staking pool's voting power in this new epoch will be updated to the total active stake.
-    /// 3. Add pending active validators to the active set if they satisfy requirements so they can vote and remove
+    /// 4. Add pending active validators to the active set if they satisfy requirements so they can vote and remove
     /// pending inactive validators so they no longer can vote.
-    /// 4. The validator's voting power in the validator set is updated to be the corresponding staking pool's voting
+    /// 5. The validator's voting power in the validator set is updated to be the corresponding staking pool's voting
     /// power.
-    public(friend) fun on_new_epoch() acquires StakePool, AptosCoinCapabilities, ValidatorConfig, ValidatorPerformance, ValidatorSet {
+    public(friend) fun on_new_epoch() acquires StakePool, AptosCoinCapabilities, ValidatorConfig, ValidatorPerformance, ValidatorSet, ValidatorFees {
         let validator_set = borrow_global_mut<ValidatorSet>(@aptos_framework);
         let config = staking_config::get();
         let validator_perf = borrow_global_mut<ValidatorPerformance>(@aptos_framework);
 
+        if (features::collect_and_distribute_gas_fees()) {
+            let validator_fees = borrow_global_mut<ValidatorFees>(@aptos_framework);
+
+            // Distribute transaction fees for each currently active validator.
+            let i = 0;
+            let len = vector::length(&validator_set.active_validators);
+            while (i < len) {
+                let validator = vector::borrow(&validator_set.active_validators, i);
+                distribute_transaction_fees(validator_fees, validator.addr);
```

Comment:
Why not just do this in update_stake_pool function?

Reply:
Then we would have to do borrow global in each update_stake_pool call? Since we need to guard `let validator_fees = borrow_global_mut<ValidatorFees>(@aptos_framework);` by `features::collect_and_distribute_gas_fees()`.

Reply:
But yes, these while loops look rather ugly...

Reply:
The main problem is that ideally we would want to wrap `ValidatorFees` in `Option`, but we cannot because it's a reference. Not to sure about the overhead of borrowing per each validator, but having a separate loop can be more optimal?